### PR TITLE
Add `ppe.nn.Ensure` 

### LINKED ISSUE #147: Module to validate a shape and dtype?
Modules like `tf.ensure_shape` can be useful in PyTorch too.
https://www.tensorflow.org/api_docs/python/tf/ensure_shape

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #147

### CODE REVIEW

**Comment by kmaehashi** [comment]
* How about supporting `None` in shape, like `ensure_shape`? I think users are likely to want to allow any size for the batch dimension.
* Maybe better having a wrapper API `ppe.ensure(tensor, *, shape=None, dtype=None, ...)`?
* Could you add tests to run these checks under jit?

**Comment by emcastillo** [comment]
> How about supporting None in shape, like ensure_shape? I think users are likely to want to allow any size for the batch dimension.

I think it is currently supported, shape should be optional.

> Maybe better having a wrapper API ppe.ensure(tensor, *, shape=None, dtype=None, ...)?

This is nice to add, but the module should be kept too so we can use it inside `torch.nn.Sequential`.

> Could you add tests to run these checks under jit?

Sure, I wonder if it will be directly exported to onnx identity. Conditionals will be ignored in the exported module.
But they should be triggered while exporting.

**Comment by kmaehashi** [comment]
> I think it is currently supported, shape should be optional.

Ah, I meant, e.g. `tf.ensure_shape(x, [None, 3])` to mark the first dimension unknown (allow any value)

**Comment by emcastillo** [comment]
seems nice :)

**Comment by emcastillo** [comment]
PTAL!

File: tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_ensure_shape.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        ((2, 4), (4,)),
+        ((2, 3, 4), (1, 4)),
+    ])
+    def test_broadcastable_shape(self, shape_t, shape_c):
+        tensor = torch.zeros(shape_t)
+        module = EnsureShapeAndDtype(shape_c, broadcastable=True)
+        module(tensor)
+
+    @pytest.mark.parametrize('shape_t, shape_c', [
+        ((3,), (2,)),
```
Comment (kmaehashi) [inline tests/pytorch_pfn_extras_tests/nn_tests/modules_tests/test_ensure_shape.py:47]: Better to test 0-dim case here?

**Review by kmaehashi — COMMENTED** [review]
LGTM except for

**Comment by emcastillo** [comment]
Addressed review comments! thanks!

File: pytorch_pfn_extras/nn/modules/ensure_shape.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+           provided one.
+       broadcastable: Check if the shapes are compatible using broadcasting
+           rules.
+       can_cast: Check if the input tensor can be casted to the provided type.
+    """
+
+    def __init__(
+            self,
+            *,
+            shape: Optional[Tuple[int]] = None,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/nn/modules/ensure_shape.py:23]: mypy failed in CI as `shape` may contain None.

```suggestion
            shape: Optional[Tuple[Optional[int]]] = None,
```

File: pytorch_pfn_extras/nn/modules/ensure_shape.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                        f' {self._dtype}')
+            else:
+                raise ValueError(
+                    f'Expected {self._dtype}, input dtype is {input.dtype}')
+        return input
+
+
+def ensure(
+        tensor: torch.Tensor,
+        shape: Optional[Tuple[int]] = None,
```
Comment (kmaehashi) [inline pytorch_pfn_extras/nn/modules/ensure_shape.py:96]:
```suggestion
        shape: Optional[Tuple[Optional[int]]] = None,
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]
LGTM except for minor comment.
Could you add new APIs to https://github.com/pfnet/pytorch-pfn-extras/blob/master/docs/source/reference/index.rst ?

**Comment by kmaehashi** [comment]
I've added an example in 8119f8f. If it's ok for you, please proceed to merge!

**Review by kmaehashi — APPROVED** [review]
